include/swift/Runtime/Metadata.h

```diff
@@ -2531,154 +2523,135 @@ struct TargetProtocolConformanceRecord {
 
 private:
   /// The protocol being conformed to.
-  RelativeIndirectablePointer<ProtocolDescriptor> Protocol;
+  ///
+  /// The remaining low bit is reserved for future use.
+  RelativeIndirectablePointerIntPair<ProtocolDescriptor, /*reserved=*/bool>
+    Protocol;
   
   // Some description of the type that conforms to the protocol.
   union {
-    /// A direct reference to the metadata.
+    /// A direct reference to the metadata, which must be uniqued before
+    /// being used.
     ///
-    /// Depending on the conformance kind, this may not be usable
-    /// metadata without being first processed by the runtime.
-    RelativeIndirectablePointer<TargetMetadata<Runtime>> DirectType;
+    /// Only valid when the \c IndirectClassOrDirectType value is
+    // \c IsDirectType.
+    RelativeDirectPointerIntPair<TargetMetadata<Runtime>,
+                                 TypeMetadataRecordKind> DirectType;
     
     /// An indirect reference to the metadata.
-    RelativeIndirectablePointer<const TargetClassMetadata<Runtime> *>
-    IndirectClass;
+    ///
+    /// Only valid when the \c IndirectClassOrDirectType value is
+    // \c IsIndirectClass.
+    RelativeDirectPointerIntPair<const TargetClassMetadata<Runtime> *,
+                                 TypeMetadataRecordKind> IndirectClass;
     
     /// The nominal type descriptor for a resilient or generic type which has
     /// instances that conform to the protocol.
-    RelativeIndirectablePointer<TargetNominalTypeDescriptor<Runtime>>
-    TypeDescriptor;
+    RelativeIndirectablePointerIntPair<TargetNominalTypeDescriptor<Runtime>,
+                                       /*always clear=*/bool> TypeDescriptor;
```

Comment:
Since we're using the spare bits to encode the type metadata kind, and we have separate kinds for direct and indirect type descriptor references, I think it might be clearer to use two union members here:

```
RelativeDirectPointerIntPair<const TargetNominalTypeDescriptor<Runtime> *,
                                 TypeMetadataRecordKind> DirectTypeDescriptor;
RelativeDirectPointerIntPair<const TargetNominalTypeDescriptor<Runtime> * const *,
                                 TypeMetadataRecordKind> IndirectTypeDescriptor;
```

since that avoids the cleverness of the `TypeMetadataRecordKind`'s kinds for nominal type descriptors just happening to line up with the way `RelativeIndirectablePointerIntPair` indicates indirection.

Reply:
Yeah, I agree that's clearer. Thanks!